### ISSUE TITLE
Troubleshooting Pytorch classifier with cleanlab

### ISSUE THREAD

**Issue by mattare2** [skorch-dev/skorch#486]
Thanks for this very helpful library! I'm trying to use my skorch-wrapped NLP classifier with the [cleanlab](https://github.com/cgnorthcutt/cleanlab) library.  My classifier works fine for training, but when I try to use the cleanlab code, I get the following error:

```
from cleanlab.classification import LearningWithNoisyLabels

# Wrap around any classifier. Yup, you can use sklearn/pyTorch/Tensorflow/FastText/etc.
lnl = LearningWithNoisyLabels(clf=net)
lnl.fit(X = np.array([f.input_ids for f in features]), s = np.array([f.label_ids for f in features]))

# Estimate the predictions you would have gotten by training with *no* label errors.
lnl.predict(val_ds.X)
```

    <ipython-input-272-eb1ed08448d3> in <module>()
    ----> 1 lnl.predict(val_ds.X)
    
    ~/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/cleanlab/classification.py in predict(self, *args, **kwargs)
        320           The test data as a feature matrix.'''
        321 
    --> 322         return self.clf.predict(*args, **kwargs)
        323 
        324 
    
    ~/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/skorch/classifier.py in predict(self, X)
        187         """
        188         y_preds = []
    --> 189         for yp in self.forward_iter(X, training=False):
        190             yp = yp[0] if isinstance(yp, tuple) else yp
        191             y_preds.append(to_numpy(yp.max(-1)[-1]))
    
    ~/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/skorch/net.py in forward_iter(self, X, training, device)
        883         iterator = self.get_iterator(dataset, training=training)
        884         for Xi, _ in iterator:
    --> 885             yp = self.evaluation_step(Xi, training=training)
        886             if isinstance(yp, tuple):
        887                 yield tuple(n.to(device) for n in yp)
    
    ~/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/skorch/net.py in evaluation_step(self, Xi, training)
        679         """
        680         with torch.set_grad_enabled(training):
    --> 681             self.module_.train(training)
        682             return self.infer(Xi)
        683 
    
    AttributeError: 'NeuralNetClassifier' object has no attribute 'module_'

Any idea where to start?

**Comment by mattare2** [comment]
Looks like any call to net.predict_proba() is causing problems...

**Comment by ottonemo** [comment]
It looks like the model is uninitialized when you are executing `lnf.predict`. Can you show us how you initialize/train the `net` instance?

**Comment by BenjaminBossan** [comment]
Yes, this seems to be the error cause.

NB: We should consider using [`check_is_fitted`](https://scikit-learn.org/stable/modules/generated/sklearn.utils.validation.check_is_fitted.html) to better catch this kind of error.

**Comment by thomasjpfan** [comment]
I am +1 on using `check_is_fitted`. sklearn uses it everywhere. :)

**Comment by BenjaminBossan** [comment]
I created a separate issue for this in #487 so that this thread can focus on the initial question.

**Comment by BenjaminBossan** [comment]
@mattare2 Did you make any progress yet?

**Comment by dreamflasher** [comment]
I am running into the same problem with the following code:

```
netclf = NeuralNetClassifier(net)
clf_isotonic = CalibratedClassifierCV(netclf, cv="prefit", method='isotonic')
clf_isotonic.fit(X, y)
```

Yields:
```
AttributeError                            Traceback (most recent call last)
<ipython-input-55-3c939d9ebf0b> in <module>
----> 9 clf_isotonic.fit(X, y)

/opt/conda/lib/python3.7/site-packages/sklearn/calibration.py in fit(self, X, y, sample_weight)
    161                 calibrated_classifier.fit(X, y, sample_weight)
    162             else:
--> 163                 calibrated_classifier.fit(X, y)
    164             self.calibrated_classifiers_.append(calibrated_classifier)
    165         else:

/opt/conda/lib/python3.7/site-packages/sklearn/calibration.py in fit(self, X, y, sample_weight)
    343         Y = label_binarize(y, self.classes_)
    344 
--> 345         df, idx_pos_class = self._preproc(X)
    346         self.calibrators_ = []
    347 

/opt/conda/lib/python3.7/site-packages/sklearn/calibration.py in _preproc(self, X)
    302                 df = df[:, np.newaxis]
    303         elif hasattr(self.base_estimator, "predict_proba"):
--> 304             df = self.base_estimator.predict_proba(X)
    305             if n_classes == 2:
    306                 df = df[:, 1:]

/opt/conda/lib/python3.7/site-packages/skorch/classifier.py in predict_proba(self, X)
    155         # Only the docstring changed from parent.
    156         # pylint: disable=useless-super-delegation
--> 157         return super().predict_proba(X)
    158 
    159     def predict(self, X):

/opt/conda/lib/python3.7/site-packages/skorch/net.py in predict_proba(self, X)
   1011         """
   1012         y_probas = []
-> 1013         for yp in self.forward_iter(X, training=False):
   1014             yp = yp[0] if isinstance(yp, tuple) else yp
   1015             y_probas.append(to_numpy(yp))

/opt/conda/lib/python3.7/site-packages/skorch/net.py in forward_iter(self, X, training, device)
    894         for data in iterator:
    895             Xi = unpack_data(data)[0]
--> 896             yp = self.evaluation_step(Xi, training=training)
    897             if isinstance(yp, tuple):
    898                 yield tuple(n.to(device) for n in yp)

/opt/conda/lib/python3.7/site-packages/skorch/net.py in evaluation_step(self, Xi, training)
    681         """
    682         with torch.set_grad_enabled(training):
--> 683             self.module_.train(training)
    684             return self.infer(Xi)
    685 

AttributeError: 'NeuralNetClassifier' object has no attribute 'module_'
```

Problem seems to be predict_proba() indeed.

**Comment by dreamflasher** [comment]
Okay, I am getting closer to a solution:

Looks like NeuralNetClassifier was not initialilzed, so there is a .initialize() missing somewhere.
```
if not netclf.initialized_:
    netclf.initialize()
```

Got me past `AttributeError: 'NeuralNetClassifier' object has no attribute 'module_'`

Now I run into: `AttributeError: 'NeuralNetClassifier' object has no attribute 'classes_'`

**Comment by BenjaminBossan** [comment]
I couldn't reproduce the first error with `CalibratedClassifierCV`, usually the net should be fitted (and hence initialized) before `predict_proba` is called.

Indeed, I got the second error. The reason, as stated, is that `NeuralNetClassifier` doesn't have the `classes_` attribute. For the reasons, look [here](https://github.com/skorch-dev/skorch/issues/465#issuecomment-489309169).

Generally, if you know the classes, you could add the attribute yourself:

```python
class MyNet(NeuralNetClassifier):
    def fit(self, X, y, **kwargs):
        self.classes_ = <code-to-get-classes>  # e.g. np.unique(y)
        return super().fit(X, y, **kwargs)
```

**Comment by githubnemo** [comment]
I think we should infer the `classes_` attribute from `y` by default and introduce an optional `classes` parameter to `NeuralNetClassifier` that, when specified, overrides the inference. This way we can disable inference (by passing `[]`), specify classes in a more complicated case (iterative training with partitioned datasets) and handle common cases automatically.

**Comment by BenjaminBossan** [comment]
If `[]` is passed, would you raise an `AttributeError` or what exactly is the idea?

**Comment by BenjaminBossan** [comment]
Fixed by #546
